src/ipa/rpi/controller/rpi/sync.cpp

```diff
@@ -119,100 +144,198 @@ void Sync::initialise()
 					state_ = State::Correcting;
-					frames = 0;						
-
-					if (!syncReady_)
-						readyCountdown_ = lastPayload_.readyFrame + frameCount_;
-
-				} else
-					break;
+				}
+
+				frames = 0;	
+	
+				if (usingWallClock_ == "yes") {
+					modeled_wall_clock_value = trendingClock.modeled_wall_clock(local.wallClock, local.sensorTimestamp, local.sequence);
+					last_wall_clock_value = lastPayload_.wallClock;
+				} else {
+					modeled_wall_clock_value = (local.sensorTimestamp) / 1000;
+					last_wall_clock_value = (lastPayload_.sensorTimestamp) / 1000;
+				}
+
+				lastPayloadFrameDuration = (lastPayload_.nextWallClock - lastPayload_.wallClock) * 1us / (lastPayload_.nextSequence - lastPayload_.sequence);
+				std::chrono::microseconds delta = (modeled_wall_clock_value * 1us) - (last_wall_clock_value * 1us);
+				unsigned int mul = (delta + lastPayloadFrameDuration / 2) / lastPayloadFrameDuration;
+				
+				delta_mod = delta - mul * lastPayloadFrameDuration;				
+
+				if (!syncReady_)
+					readyCountdown_ = lastPayload_.readyFrame + frameCount_;
+				
+				if (!syncReady_ && lastPayload_.readyFrame) {
+					expected = lastPayload_.wallClock * 1us + lastPayload_.readyFrame * lastPayloadFrameDuration;
+				}
+			} else {
+				
+				break;}
+		}
+		if (syncReady_ && !frames) {
+			delta_mod = trendig_error.trending_error(last_wall_clock_value * 1us, modeled_wall_clock_value * 1us, lastPayloadFrameDuration, local.sequence);
+			if (abs(delta_mod) > 50us) {
+				trendig_error.updating_values(delta_mod);
+				state_ = State::Correcting;
+			}
 		}
 
-		std::chrono::microseconds lastPayloadFrameDuration = (lastPayload_.nextWallClock - lastPayload_.wallClock) * 1us / (lastPayload_.nextSequence - lastPayload_.sequence);
-		std::chrono::microseconds delta = (local.wallClock * 1us) - ((lastPayload_.wallClock * 1us) + frames * lastPayloadFrameDuration);
-		unsigned int mul = (delta + lastPayloadFrameDuration / 2) / lastPayloadFrameDuration;
-		std::chrono::microseconds delta_mod = delta - mul * lastPayloadFrameDuration;
-		LOG(RPiSync, Info) << "Current frame : seq " << local.sequence << " ts " << local.wallClock << "us"
-				   << " frame offset " << frames << " est delta " << delta << " (mod) " << delta_mod << " ready countdown " << readyCountdown_;
+		if (state_ == State::Correcting) {
+			LOG(RPiSync, Info) << "Correcting by " << delta_mod;
 
-		if (state_ == State::Correcting)  {
 			status.frameDurationOffset = delta_mod;
 			state_ = State::Stabilising;
-			LOG(RPiSync, Info) << "Correcting offset " << status.frameDurationOffset;
-
 		} else if (state_ == State::Stabilising) {
-			status.frameDurationOffset = 0s;
-			LOG(RPiSync, Info) << "Stabilising duration ";		
+
+			status.frameDurationOffset = 0s;		
 			state_ = State::Idle;
 		}
 
-		if (!syncReady_ && readyCountdown_ && !(readyCountdown_ - frameCount_)) {
+		if (!syncReady_ && readyCountdown_ && local.wallClock * 1us > expected - lastPayloadFrameDuration / 2 && expected != 0us) {
```

Comment:
Do we need `readyCountdown_` here? Presumably once we've hit the right time (`expected`), then that's all we care about? I wonder slightly whether `readyCountdown_` could in fact be removed completely.